### PR TITLE
Add support for skipping the next translation group

### PR DESCRIPTION
**Pull request by dyoo**
This is the beginning of support for handling special comments for translation, https://github.com/google/mdbook-i18n-helpers/issues/63

This adds support for adding a comment of the form: `<!-- mdbook-xgettext: skip -->`.  This will cause the system to skip the next message group that would otherwise be translated.

It adds a dependency to the regex crate to match for the comment skip pattern.

### CODE REVIEW

**Review by djmitche — APPROVED** [review]
The implementation looks solid. However, I suspect that `mdbook-xgettext: skip` is rather a lot to remember when writing a slide.

Thoughts on using `<!-- NO-TRANSLATE -->` instead? I think the use of all-caps signifies that this comment is for machine interpretation, and the words are specific enough that they would not conflict with anything not translation-related, so there is no need to scope them with `mdbook-xgettext`.

File: src/lib.rs [hunk 1]
```diff
@@ -951,4 +1001,94 @@ BOB
...
+def f(x): return x * x
+```
+This should be translated.
+"#,
+            vec![(5, "This should be translated.")],
+        );
+    }
+
+    #[test]
+    fn extract_messages_skip_back_to_back() {
```
[the remark this answers is not in the document]
Comment (mgeisler) [reply src/lib.rs:1080]: > Could you add another test which shows what happens when the comment is in the middle of a line?

Related to this, what happens with `foo <!-- mdbook-xgettext: skip --> **bold** bar`? Do we skip `**bold**` but still form a single paragraph? Dont' worry too much if the output is weird in these corner-cases: I'm mostly interested in documenting them via a test. Speaking of documentation... you should add a section to `USAGE.md` about this new feature.

I would also be interested in seeing a test that shows skipping the second bullet here:

> - Not skipped
> - Skipped
> - Also not skipped

That's about the most complex markup we use in Comprehensive Rust.

**Comment by mgeisler** [comment]
> Thoughts on using `<!-- NO-TRANSLATE -->` instead? I think the use of all-caps signifies that this comment is for machine interpretation, and the words are specific enough that they would not conflict with anything not translation-related, so there is no need to scope them with `mdbook-xgettext`.

Yeah, it's rather long indeed... I guess I was thinking people would copy-paste in front of large code blocks as needed.

I suggested the long form in #63 and it's slightly inspired by how `xgettext` supports special [`xgettext:c-format` and `xgettext:no-c-format` comments](https://www.gnu.org/software/gettext/manual/gettext.html#c_002dformat-Flag). However, the manual also talks about a [`TRANSLATORS:` prefix](https://www.gnu.org/software/gettext/manual/gettext.html#Names), which is more your style (I believe you trigger this with `xgettext --add-comments=TRANSLATORS`).

I'm not entirely sure where we should land, but we should try to think of how skipping can work with more directives in the future: translator comments and a message context.

**Comment by dyoo** [comment]
@mgeisler 's questions on what happens with the current implementation and inline sections made me reconsider.  I didn't know offhand what exactly would happen, and that to me is a warning sign.

I'd rather it be obviously clear.  So maybe it might make a safer and robust feature to have the skips define **regions**, rather than implicitly skip the next block.  Explicitly marking off the start and end of a skip section feels both simpler and clearer to me.

It would look something like this

```
<!-- mdbook-xgettext: skip-start -->
This content is ignored.
<!-- mdbook-xgettext: skip-end -->
```

This would make it easier to see what sections are being ignored, removing ambiguity on the boundaries of the skipped region.

Would you be amendable to this proposed change to the feature?

**Comment by mgeisler** [comment]
> It would look something like this
> 
> ```
> 
> This content is ignored.
> 
> ```
> 
> This would make it easier to see what sections are being ignored, removing ambiguity on the boundaries of the skipped region.
> 
> Would you be amendable to this proposed change to the feature?

I'm more thinking that we could have both. The immediate use case for skipping is to remove some very large code listings from the PO files for Comprehensive Rust. There your current implementation will already work great and I think we should merge it already.

A next step could be to support skipping regions like you suggest — I think this is a great idea! Infact, I just found a tool today which supports similar comments: [mdpo](https://mondeja.github.io/mdpo/latest/commands.html#disabling-extraction).

Seeing the comments they support gives me an idea: we could use `i18n` as a shorthand for `mdbook-i18n-helpers` and implement these comments:

- `i18n-skip`: this PR, skips next event,
- `i18n-disable`: starts skipping events,
- `i18n-enable`: stops skipping,
- `i18n-comment`: translator comment,
- `i18n-context` and `i18n-ctx`: sets message context for next (or following?) event.

File: src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (mgeisler) [reply src/lib.rs:1080]: I suggest we merge this after adding a few more interesting tests. We can then brainstorm the keyword a bit and extend with extra functionality as needed.

**Comment by dyoo** [comment]
Sounds good!  I have added more tests but have left the rest of the implementation as is; we can follow up with changing the keyword in a subsequent PR.

File: src/lib.rs [hunk 2]
```diff
@@ -207,13 +222,14 @@ pub fn group_events<'a>(events: &'a [(usize, Event<'a>)]) -> Vec<Group<'a>> {
             // make the group self-contained.
             Event::Start(Tag::Paragraph | Tag::CodeBlock(..)) => {
                 // A translatable group starts here.
-                groups.push(state.into_group(idx, events));
+                groups.push(state.into_group(idx, events, &mut skip_next_group));
+
                 state = State::Translate(idx);
```
Comment (mgeisler) [inline src/lib.rs:227]: I would like to avoid the `&mut bool` in-out parameter in `into_group`. I find that in-out parameters makes it hard to know what is updated where (and why it is updated).

I think we could avoid it if we instead change `state` to `State::Skip` here when `skip_next_group == true`. Perhaps we could have a pure function which takes a `State` and a Boolean and returns the new state and the new Boolean?

Comment (dyoo) [reply src/lib.rs:248]: Sounds good to me.  I'll add that in.  I had done it that way initially, and started getting tired of the state threading.  But given that we'll eventually expand what we're keeping track of to include translator comments and message contexts, we might as well prepare the way for it.

I've adjusted the code to manage a GroupingContext object, and have into_group() consume and produce it.

**Comment by dyoo** [comment]
I note that there is appears to be a bug in the Markdown parser that I've reported upstream as https://github.com/raphlinus/pulldown-cmark/issues/712.  The bug causes content on the same line as an inline HTML comment to be "glued" to the comment.  In the case of our comment-skipping directive, this means that some of the behavior in this PR is due to  *accident* rather than intention.  :)

I do have a workaround for the inline HTML comment bug, prepared in https://github.com/google/mdbook-i18n-helpers/compare/main...dyoo:mdbook-i18n-helpers:isolate_first_html_comment.  If it turns out that fixing the bug upstream is taking too much time, I can clean up that branch and prepare it for a PR.  I wanted to give you a heads up so that we're aware of the issue.

**Comment by mgeisler** [comment]
Hey @dyoo, thanks for the update! I ran `dprint fmt` (the formatter we use in this repository) and pushed the result back to your branch. I also squashed the commits: that way every commit will pass all CI tests.

In general, I prefer to edit existing commits instead of having a string of commits in the history — especially if the string of commits represent fixes to comments that came up through code review. I wrote down some thoughts on this [more than 10 years ago](https://docs.zerovm.org/contributing.html), in case you're curious :smile:
